ae/declare static

archie-mckenzie opened the pull request:
# What
Added support such that using `declareStatic` is allowed in call expressions for both babel and swc

# Why
We've recently added Static as a new powerful feature described [here](https://generaltranslation.com/en-US/blog/gt-next_v6_8_0)
This adds logic to our compiler to ensure its proper usage

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Add end-to-end support for declareStatic in translation calls/components, updating parsers/validators, SWC rules, and exporting declareStatic across packages.
> 
> - **Static Analysis / Parsers**:
>   - CLI: Add `parseDeclareStatic` and extend string parsing to resolve `declareStatic(...)` (incl. `await`, cross-file imports, re-exports, aliasing), plus new concise warnings and code formatting.
>   - JSX Parser: Improve import aliasing and cross-file resolution; ensure local defs shadow imports.
>   - Tests: Add extensive cases for re-exports, aliasing, multi-level chains, and shadowing.
> - **Compiler (Babel)**:
>   - Move and expand `validateUseGTCallback` to accept string or `declareStatic(...)`; integrate with processing/injection flows; add comprehensive unit tests.
> - **SWC Plugin**:
>   - Allow only string literals or `declareStatic(...)` (incl. `await`) in translation args; add `declareStatic` detection/validation and tests; track as GT import.
> - **API Surface**:
>   - Export `declareStatic` alongside `Static` in `react-core`, `react`, and `next` (client/server/types); update docs/comments.
> - **Misc**:
>   - Update GT library upstream compatibility (react/native cross-compat).
>   - ESLint: ignore `_`-prefixed unused variables and vars.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e6264517c090ada13afa66992e8a0ed779469a4c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

archie-mckenzie committed:
declare static

archie-mckenzie committed:
ran lint:fix

archie-mckenzie committed:
updated jsdoc comments

pie575 committed:
added declareStatic support from babel

pie575 committed:
added tests for declareStatic functions

pie575 committed:
working Rust compiler with declareStatic

pie575 committed:
fix tests

pie575 committed:
added await logic, started building cli

pie575 committed:
Merge branch 'main' into ae/declare-static

pie575 committed:
working CLI for parsing declare static

pie575 committed:
Merge branch 'main' into ae/declare-static

eoinest commented on packages/cli/src/console/index.ts:
```diff
@@ -234,6 +234,19 @@ export const warnRecursiveFunctionCallSync = (
+): string =>
+  withLocation(
+    file,
+    withStaticError(
```
Right now `withStaticError()` is explicitly for `<Static>`, not `declareStatic()`. Either update `withStaticError()`'s message to be more general or make a different helper function

eoinest commented on packages/cli/src/console/index.ts:
```diff
@@ -234,6 +234,19 @@ export const warnRecursiveFunctionCallSync = (
+  withLocation(
+    file,
+    withStaticError(
+      `Could not resolve ${colorizeFunctionName(functionName)}. This call is either not wrapped in declareStatic() or uses circular/looped file imports. Ensure the function is properly wrapped with declareStatic() and does not have circular import dependencies.`
```
You actually have three errors going on here: (1) not wrapped in declare static (2) circular imports (3) recursive calls. While (2) and (3) are similar enough to be its own message, option (1) should definitely be its own distinct error message.

eoinest commented on packages/cli/src/react/jsx/utils/jsxParsing/parseJsx.ts:
```diff
@@ -750,7 +750,7 @@ function resolveStaticComponentChildren({
 }
 
-function resolveStaticFunctionInvocationFromBinding({
+export function resolveStaticFunctionInvocationFromBinding({
```
i might be wrong, but doesnt seem like you're using this function outside of this file, please do not export

eoinest commented on packages/cli/src/react/jsx/utils/parseDeclareStatic.ts:
```diff
@@ -0,0 +1,625 @@
+import * as t from '@babel/types';
```
move parseDeclareStatic and types.ts into a declareStaticParsing directory.

I would say rule of thumb is when you have multiple files that are only meant to do one thing, they get their own dir.

this is also helpful b/c this shows type.ts contains parseDeclareStatic content only

eoinest commented on packages/cli/src/react/jsx/utils/parseDeclareStatic.ts:
```diff
@@ -0,0 +1,625 @@
+const traverse = traverseModule.default || traverseModule;
+
+// Nested arrays of strings (deprecated - kept for backwards compatibility)
+export type StringTree = (string | StringTree)[];
```
can this be moved to type.ts. Furthermore, anything that is deprecated should be marked as follows:
```js
/** @deprecated use ... instead */
export type StringTree = (string | StringTree)[];
```

Furthermore, why do we need backwards compatibility here? Backwards compatibility should really only concern user-facing elements. This is pretty deep inside of the codebase, so I am unsure how backwards compatibility is relevant here.

eoinest commented on packages/cli/src/react/jsx/utils/parseDeclareStatic.ts:
```js
// example
gt('hello' + declareStatic())
```
```
//output
┌  Generating source templates...
│
▲  CLI tool encountered 1 warnings while scanning for translatable content.
│  • Warning: /Users/ernestmccarter/Documents/dev/playground/test-declare-static/app/page.tsx (9:15): <Static> rules violation: Could not resolve declareStatic. This call is either not wrapped in declareStatic() or uses circular/looped file imports. Ensure the function is properly wrapped with declareStatic() and does not have circular import dependencies.
│
■  Error: CLI tool encountered 1 syntax errors while scanning for translatable content. To ignore these errors, re-run with --ignore-errors
│  • /Users/ernestmccarter/Documents/dev/playground/test-declare-static/app/page.tsx (9:5): Found non-string literal ('hello' + declareStatic()). Change the value to a string literal to ensure this content is translated.
```

A few things about this behavior: (1) the error message should be more descriptive. The issue is that there is no function invocation inside of `declareStatic()`, not that `declareStatic()` could not be resolved. (2) There is both a warning and an error. There should only be one message. (3) All `declareStatic()` violations should probably have the same prefix such as `declareStatic()` rule violation.

eoinest commented on packages/cli/src/react/jsx/utils/parseDeclareStatic.ts:
Ive talked with Archie. He says that we should change `declareStatic()` violations from warnings to errors. This is to stop ppl from shooting themselves in the foot as much as possible.

eoinest commented on packages/cli/src/react/jsx/utils/parseDeclareStatic.ts:
I notice a lot of functions in this file are being exported, but are not used outside of this file. Please do not export these.

If you want to export a particular function to test it follow this standard:
```js
// Entry point functions get exported inline
export function entryPointFunction() { ... }

// Helper functions can be exported for testing, but should be done with something denoting them as being exported for testing purposes
function helperFunction() { ... }

// Unlike entry point functions helper function exports for testing are not inlined so readers don't get confused
export { helperFunction as _helperFunction };
```

eoinest commented on packages/cli/src/react/jsx/utils/parseDeclareStatic.ts:
Something is going wrong with how you are doing import tracking and using bindings when there are two possible bindings: an in file binding, an imported binding.
```tsx
//page.tsx
import { useGT, declareStatic } from 'gt-next'
import { fn1 } from '@/lib/utils1'

export default function Home() {
  const gt = useGT();
  const fn1 = () => {
    return 'internal'
  }
  gt(declareStatic(fn1()))
  return null;
}

// /lib/utils1.ts
export function fn1() {
  return 'external'
}
```
```
// output
▲  CLI tool encountered 1 warnings while scanning for translatable content.
│  • Warning: /Users/ernestmccarter/Documents/dev/playground/test-declare-static/app/page.tsx (10:5): <Static> rules violation: Could not resolve declareStatic. This call is either not wrapped in declareStatic() or uses circular/looped file imports. Ensure the function is properly wrapped with declareStatic() and does not have circular import dependencies.
│
■  Error: CLI tool encountered 1 syntax errors while scanning for translatable content. To ignore these errors, re-run with --ignore-errors
│  • /Users/ernestmccarter/Documents/dev/playground/test-declare-static/app/page.tsx (10:5): Found non-string literal (declareStatic(fn1())). Change the value to a string literal to ensure this content is translated.
```

My advice is two fold:
(1) (more likely) you might be preferring imports over local function definitions. Check out the behavior 282-348. I think there is something going wrong here.
(2) (less likely) pay close attention to how you are passing around the `NodePath` argument. It seems that you are just prop drilling the variable. This technically works, but it doesn't always do scoping correctly. Remember `NodePath` is a wrapper around a `Node` object. If you want to get the `NodePath` for a child `Node` you can call `nodePath.get('childname'). For example:
```js
const returnPathNode: NodePath<t.ReturnStatement> = ...;
// get the "node" for a return statement's argument field:
const argumentNode: t.Expression = returnPathNode.node.argument;
// get the "NodePath" for a return statement's argument field:
const argumentNodePath: NodePath<t.Expression> = returnPath.get('argument');
```
This is IMPORTANT because it could possibly change the scope eg `returnPathNode.scope.getBinding('example')` might be different from `argumentNodePath.scope.getBinding('example')`.

eoinest commented on packages/cli/src/react/jsx/utils/parseDeclareStatic.ts:
Looks like you don't have support for re-exports. These come in a few different flavors, but it's important to nail these because these are pretty common in large code bases. I'll give you two examples to help.


```ts
// utils1.ts
export function fn1() {
  return 'external'
}
```

**Example 1: namespace re-exports:**
```tsx
// page.tsx
import { useGT, declareStatic } from 'gt-next'
import { fn1 } from '@/lib/utils2'  // <------

export default function Home() {
  const gt = useGT();
  gt(declareStatic(fn1()))
  return null;
}
```
```ts
// utils2.ts
export * from './utils1'
```

**Example 2: named re-exports:**
```tsx
// page.tsx
import { useGT, declareStatic } from 'gt-next'
import { fn1 } from '@/lib/utils3' // <------

export default function Home() {
  const gt = useGT();
  gt(declareStatic(fn1()))
  return null;
}
```
```ts
// utils3.ts
export { fn1 } from './utils1'
```

Example 1 is a bit more tricky. Because we don't know what exactly is being exported, we need to check every namespace export. For example, `utils2.ts` could look something like this. We have to search each file for the `fn1` export.
```ts
export * from './utils1'
export * from './utils4'
export * from './utils5'
export * from './utils6'
```

eoinest commented on packages/cli/src/react/jsx/utils/parseDeclareStatic.ts:
```diff
@@ -0,0 +1,625 @@
+ */
+export function handleStaticExpression(
+  expr: t.Expression,
+  tPath: NodePath,
```
Optional, but consider adding strict typing for the `tPath` parameter. You can add typing for `NodePath`s by passing a type generic: `NodePath<t.Expression>`.

eoinest commented on packages/compiler/src/processing/first-pass/processCallExpression.ts:
```diff
@@ -135,7 +135,8 @@ function handleUseGTCallback(
     state,
-    useGTCallbackParams.content!,
+    useGTCallbackParams.content ?? '', // content will only be undefined if it is a declareStatic function invocation
+    // Todo: update this logic to store a flag to skip injection on the second pass
```
todo should be capitalized: `TODO:`

pie575 committed:
addressed comments

pie575 committed:
merge with main

pie575 committed:
Revert "merge with main"

This reverts commit 4f9196bcc3a109f0631a76ac93170cb5d9a23f70.

ErnestM1234 committed:
chore: merge main

ErnestM1234 committed:
fix: greptile

eoinest commented on packages/compiler/src/transform/validation/validateTranslationFunctionCallback.ts:
```diff
@@ -0,0 +1,289 @@
+
+  // Fallthrough: expression type not supported (e.g., plain identifiers/variables)
+  errors.push('Variables are not allowed');
+  return { errors };
```
TODO

ErnestM1234 committed:
fix: merge conflicts

eoinest commented on packages/compiler/src/transform/validation/validateTranslationFunctionCallback.ts:
```diff
@@ -0,0 +1,289 @@
+    return {
+      errors,
+    };
+  }
```
TODO

eoinest commented on packages/cli/src/react/jsx/utils/parseString.ts:
```diff
@@ -0,0 +1,728 @@
+        );
+      }
+      return null;
+    }
```
Known issue, todo

ErnestM1234 committed:
fix: greptile

ErnestM1234 committed:
fix: compiler

ErnestM1234 committed:
fix: allow Static to accept more behavior as direct children (#874)

* fix: allow Static to accept more behavior as direct children

* fix: static support for arbitrary children

* test: branching and other static children

* fix: consolelog

* fix: remove comments

* fix: imports

ErnestM1234 committed:
fix: add a test to make sure we are not restricting contents of m() function

pie575 committed:
fixed msg/m cases with ignoreDynamicContent

pie575 committed:
parseStringFunction test

ErnestM1234 committed:
feat: add declareVar() (#882)

# Why
libraries need a `declareVar()` + helper functions
# What
`declareVar()` + helpers

ErnestM1234 committed:
fix: intl parse resolution

ErnestM1234 committed:
fix: only resolve imported declareStatic(), not just any function named 'declareStatic'

ErnestM1234 committed:
chore: merge main

ErnestM1234 committed:
fix: rollup build warning

ErnestM1234 committed:
fix: react core tests

ErnestM1234 committed:
feat: add declareVar parsing

ErnestM1234 committed:
fix: declar var support

ErnestM1234 committed:
fix: unclear error message

ErnestM1234 committed:
fix: declareStatic argument validation

ErnestM1234 committed:
chore: changeset

ErnestM1234 committed:
chore: alphas

ErnestM1234 committed:
chore: gen version

ErnestM1234 committed:
fix: useGT validation

eoinest commented on packages/cli/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "gtx-cli",
-  "version": "2.5.33",
+  "version": "2.5.34-alpha.0",
```
revert

ErnestM1234 committed:
chore: bump alpha

ErnestM1234 committed:
chore: restore verions

ErnestM1234 committed:
chore: merge main

ErnestM1234 committed:
fix: changeset
